<div>
<img src="https://i.ibb.co/v3CvVz9/udd-short.png" width="150"/>
    <br>
    <strong>Universidad del Desarrollo</strong><br>
    <em>Magíster en Data Science</em><br>
    <em>Profesora: Maria Paz Raveau</em><br>
    <em>Asignatura: Procesamiento de Lenguaje Natural</em><br>

</div>

# **Tarea 1**

*Fecha de Entrega: Domingo 20, Julio 2025.*

**Nombre Estudiante**: Victor Saldivia Vera

### **01. Intrucciones y Enunciado**

Suponga que ud. es contactado por alguna agencia gubernamental que quiere saber qué dijeron los participantes de los cabildos de 2016. Ud. debe entonces, basándose en los textos de fundamento, responder a este requerimiento.  Use para ello la base de datos que vimos en clases, de los cabildos provinciales. 

- Escoja alguna de los conceptos que vienen en la base de datos (evite las que no tengan muchas entradas). Hemos visto cómo generar WordClouds con tokens y con bigramas. *¿Es posible valerse de las etiquetas gramaticales para mejorar el resultado?* 
- Pruebe haciendo un WordCloud con tokens/sustantivos,  y otro con  bigramas/sustantivo-adjetivo. Recuerde pre-procesar correctamente.
- Con estos mismo bigramas/sustantivo-adjetivo, haga una red de bigramas.
- Ayúdese con frases de ejemplo, y de los WordClouds, para escribir una descripción del concepto.


### **02. Descripción de la Data**

Carga y primeras exploraciones del DataFrame original.  
- Columnas disponibles: 12  
- Tamaño del dataset: 30414 registros  
- Conteo de nulos en las columnas de texto: 0

In [105]:
import pandas as pd
df = pd.read_excel('data/resultadocabildoprovincial.xlsx')
print(f'Dimensión: {df.shape}'), 
print(f'Nombre de Columnas: {df.columns}')

Dimensión: (30414, 12)
Nombre de Columnas: Index(['region_nombre', 'provincia_nombre', 'comuna_nombre', 'cabildo_id',
       'cabildo_nombre', 'cabildo_direccion', 'num_mesa', 'pregunta_glosa',
       'concepto_nombre', 'concepto_otro', 'tipo-eleccion', 'fundamento'],
      dtype='object')


### **03. Librerías a Utilizar**
Importación de librerías necesarias:

- **Pandas**, **NumPy** y expresiones regulares para limpieza  
- **Stanza** para NLP (tokenización, POS, lematización)  
- **NLTK** para stopwords y collocations  
- **WordCloud**, **NetworkX** y **Matplotlib** para visualización  
- **tqdm** para barras de progreso.

In [106]:
import numpy as np
import re
import os
import unicodedata
from collections import Counter

# Librerías de Procesamiento de Texto
import nltk
from nltk.corpus import stopwords
from nltk.collocations import BigramCollocationFinder, BigramAssocMeasures
import stanza

# Visualización
import matplotlib.pyplot as plt
from matplotlib.patches import Patch
from wordcloud import WordCloud
import networkx as nx

# La librería de warnings se utiliza para manejar las advertencias
import warnings
warnings.filterwarnings('ignore')

# Paquete muestra Barra de Progreso
from tqdm import tqdm
tqdm.pandas()

Se estableció `KMP_DUPLICATE_LIB_OK=TRUE` para evitar conflicto de múltiples runtimes OpenMP en Windows. Es un workaround temporal *(solución temporal)*.

In [107]:
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"

In [ ]:
# Descargas (Nota: Ejecutar una sola vez en entorno local)
# nltk.download('punkt')
# nltk.download('stopwords')
# stanza.download('es')

#### Configuración de Stanza para el NLP

In [108]:
nlp = stanza.Pipeline('es', processors='tokenize,pos,lemma', use_gpu=False)  # Cambiar a True si hay GPU


2025-07-20 18:43:36 INFO: Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES
2025-07-20 18:43:36 INFO: Downloaded file to C:\Users\Victo\stanza_resources\resources.json
2025-07-20 18:43:36 WARNING: Language es package default expects mwt, which has been added
2025-07-20 18:43:37 INFO: Loading these models for language: es (Spanish):
| Processor | Package           |
---------------------------------
| tokenize  | combined          |
| mwt       | combined          |
| pos       | combined_charlm   |
| lemma     | combined_nocharlm |

2025-07-20 18:43:37 INFO: Using device: cpu
2025-07-20 18:43:37 INFO: Loading: tokenize
2025-07-20 18:43:37 INFO: Loading: mwt
2025-07-20 18:43:37 INFO: Loading: pos
2025-07-20 18:43:39 INFO: Loading: lemma
2025-07-20 18:43:40 INFO: Done loading processors!


### **04. Carga y Exploración Inicial**


In [109]:
df.head()

region_nombre provincia_nombre comuna_nombre  cabildo_id  \
0  Arica y Parinacota            Arica         Arica          81   
1  Arica y Parinacota            Arica         Arica          81   
2  Arica y Parinacota            Arica         Arica          81   
3  Arica y Parinacota            Arica         Arica          81   
4  Arica y Parinacota            Arica         Arica          81   

       cabildo_nombre        cabildo_direccion  num_mesa  \
0  Colegio San Marcos  Santa María 2140, Arica         1   
1  Colegio San Marcos  Santa María 2140, Arica         1   
2  Colegio San Marcos  Santa María 2140, Arica         1   
3  Colegio San Marcos  Santa María 2140, Arica         1   
4  Colegio San Marcos  Santa María 2140, Arica         1   

                                      pregunta_glosa  \
0  ¿Cuáles son los valores y principios más impor...   
1  ¿Cuáles son los valores y principios más impor...   
2  ¿Cuáles son los valores y principios más impor...   
3  ¿Cuáles son los valores y principios más impor...   
4  ¿Cuáles son los valores y principios más impor...   

                                     concepto_nombre concepto_otro  \
0                                 Autonomía/libertad           NaN   
1                                         Democracia           NaN   
2                                  Descentralización           NaN   
3                                  Equidad de género           NaN   
4  Respeto/conservación de la naturaleza o medio ...           NaN   

  tipo-eleccion                                         fundamento  
0             A                                     sin fundamento  
1             A                                     Sin fundamento  
2             A  La descentralizacion de ser autonoma por ser z...  
3             A                                     Sin fundamento  
4             A                                     Sin fundamento

#### Conteo de nulos en columnas de texto


In [110]:
texto_col = "fundamento"
print(df[texto_col].isna().sum())

0


#### Estadísticas Descriptivas sobre el texto


In [111]:
df['len_chars'] = df[texto_col].str.len() # columna de longitud de caracteres
df['len_tokens_raw'] = df[texto_col].str.split().str.len() # columna de tokens sin procesar
df[['len_chars','len_tokens_raw']].describe()

len_chars  len_tokens_raw
count  30414.000000    30414.000000
mean      69.090123       10.828270
std       69.670394       10.993367
min        1.000000        1.000000
25%       15.000000        2.000000
50%       47.000000        7.000000
75%      102.000000       16.000000
max      300.000000       61.000000

#### Eliminación de nulos y duplicados


In [112]:
before = len(df)
df = df.dropna(subset=[texto_col])
df = df.drop_duplicates(subset=[texto_col])
print(f"Filas eliminadas: {before - len(df)}")

Filas eliminadas: 11088


### **05. Selección de Concepto**
Se exploran las frecuencias de `concepto_nombre` y se elige el concepto.

- Se elige el concepto `Democracia` ya que tienen una alta frecuencia (Top 3) 
- Se construye yb sub conjunto `subset` filtrado por el concepto elegido

In [113]:
# Distribución de frecuencias para elegir uno con suficientes observaciones
concepto_col = "concepto_nombre" 
concept_counts = df[concepto_col].value_counts()
concept_counts.head(20)

concepto_nombre
Otro                                                      3976
Igualdad                                                   519
Democracia                                                 507
Descentralización                                          503
Justicia                                                   492
Respeto/conservación de la naturaleza o medio ambiente     484
A la vida                                                  440
A la salud                                                 437
A la educación                                             428
Respeto                                                    385
Fuerzas Armadas                                            381
Bien Común/comunidad                                       381
Congreso o parlamento (estructura y funciones)             373
A la vivienda digna                                        365
Plebiscitos, referendos y consultas                        361
Respeto por la constitución            

#### Selección de Concepto

In [114]:
concepto_elegido = "Democracia" 

#### Creación del Subconjunto filtrado
Se elige el concepto 'Democracia' el cual tiene una frecuencia de (507 ocurrencias) para garantizar masa textual suficiente y diversidad léxica. Se descartaron conceptos con baja frecuencia para evitar resultados inestables en nubes de palabras y bigramas.”



In [115]:
subset = df[df[concepto_col] == concepto_elegido].copy()
print("Filas del concepto:", len(subset))

Filas del concepto: 507


In [116]:
# Subconjunto de datos para el concepto elegido
subset.head(10)

region_nombre provincia_nombre  comuna_nombre  cabildo_id  \
1    Arica y Parinacota            Arica          Arica          81   
142  Arica y Parinacota            Arica          Arica          81   
169  Arica y Parinacota            Arica          Arica          81   
283  Arica y Parinacota            Arica          Arica          81   
316  Arica y Parinacota            Arica          Arica          81   
363  Arica y Parinacota            Arica          Arica          81   
491  Arica y Parinacota            Arica          Arica          81   
526  Arica y Parinacota            Arica          Arica          81   
746  Arica y Parinacota       Parinacota          Putre         121   
945            Tarapacá          Iquique  Alto Hospicio         107   

                cabildo_nombre              cabildo_direccion  num_mesa  \
1           Colegio San Marcos        Santa María 2140, Arica         1   
142         Colegio San Marcos        Santa María 2140, Arica         5   
169         Colegio San Marcos        Santa María 2140, Arica         6   
283         Colegio San Marcos        Santa María 2140, Arica         9   
316         Colegio San Marcos        Santa María 2140, Arica        10   
363         Colegio San Marcos        Santa María 2140, Arica        11   
491         Colegio San Marcos        Santa María 2140, Arica        14   
526         Colegio San Marcos        Santa María 2140, Arica        15   
746  Liceo Granaderos de Putre  Teniente del Campo 242, Putre         2   
945     San Antonio de Matilla  Los Kiwis 3443, Alto Hospicio         2   

                                        pregunta_glosa concepto_nombre  \
1    ¿Cuáles son los valores y principios más impor...      Democracia   
142  ¿Cuáles son los valores y principios más impor...      Democracia   
169  ¿Cuáles son los valores y principios más impor...      Democracia   
283  ¿Cuáles son los valores y principios más impor...      Democracia   
316  ¿Cuáles son los valores y principios más impor...      Democracia   
363  ¿Cuáles son los valores y principios más impor...      Democracia   
491  ¿Cuáles son los valores y principios más impor...      Democracia   
526  ¿Cuáles son los valores y principios más impor...      Democracia   
746  ¿Cuáles son los valores y principios más impor...      Democracia   
945  ¿Cuáles son los valores y principios más impor...      Democracia   

    concepto_otro tipo-eleccion  \
1             NaN             A   
142           NaN             A   
169           NaN             P   
283           NaN             P   
316           NaN             A   
363           NaN             A   
491           NaN             A   
526           NaN             A   
746           NaN             A   
945           NaN             A   

                                            fundamento  len_chars  \
1                                       Sin fundamento         14   
142  Que el ciudadano pueda participar en la toma d...        141   
169  Las condiciones actuales no constituyen democr...        118   
283  por la estructura establecida en el dialogo an...         52   
316  debemos pasar de una democracia representativa...         66   
363                               por eso estamos aqui         20   
491  Se garantice una democracia vinculante que se ...         89   
526                    la base y el principio de todos         31   
746  Mejorar  en todos los aspectos los sistemas de...        258   
945  Elemento fundamental en la nueva constitución ...         54   

     len_tokens_raw  
1                 2  
142              22  
169              17  
283               8  
316               9  
363               4  
491              14  
526               7  
746              41  
945               8

### **06. Pre-Procesamiento de Texto**

**Limpieza de texto**  
1. Remoción de Ruido Semántico (ejemplo: “sin fundamento”)  
2. Normalización a minúsculas
3. Eliminación de caracteres especiales y tildes
4. Eliminación de espacios extras

In [120]:
PLACEHOLDER_PATTERNS = {
    'sin fundamento','sin  fundamento','sinfundamento','sf','nuevo concepto'
}

def is_placeholder(text):
    if not isinstance(text, str):
        return True
    t = text.strip().lower()
    t = re.sub(r'\s+', ' ', t)
    return t in PLACEHOLDER_PATTERNS

before_ph = len(df)
df = df[~df[texto_col].apply(is_placeholder)]
print(f"{before_ph - len(df)} filas eliminadas por ser placeholders")

0 filas eliminadas por ser placeholders


#### Normalización de Texto

In [121]:
def normalize_text(text: str) -> str:
    text = str(text).lower() #
    text = unicodedata.normalize('NFKD', text).encode('ascii','ignore').decode('utf-8','ignore')  # eliminar tildes
    text = re.sub(r'[^a-z0-9\\s]', ' ', text) # eliminar caracteres especiales
    text = re.sub(r'\\s+', ' ', text).strip() # eliminar espacios extra
    return text

subset['texto_norm'] = subset[texto_col].apply(normalize_text)

In [122]:
subset

region_nombre  provincia_nombre comuna_nombre  \
1                        Arica y Parinacota             Arica         Arica   
142                      Arica y Parinacota             Arica         Arica   
169                      Arica y Parinacota             Arica         Arica   
283                      Arica y Parinacota             Arica         Arica   
316                      Arica y Parinacota             Arica         Arica   
...                                     ...               ...           ...   
30184  Magallanes y de la Antártica Chilena  Última Esperanza       Natales   
30212  Magallanes y de la Antártica Chilena  Última Esperanza       Natales   
30251  Magallanes y de la Antártica Chilena  Última Esperanza       Natales   
30290  Magallanes y de la Antártica Chilena  Última Esperanza       Natales   
30372  Magallanes y de la Antártica Chilena  Última Esperanza       Natales   

       cabildo_id            cabildo_nombre               cabildo_direccion  \
1              81        Colegio San Marcos         Santa María 2140, Arica   
142            81        Colegio San Marcos         Santa María 2140, Arica   
169            81        Colegio San Marcos         Santa María 2140, Arica   
283            81        Colegio San Marcos         Santa María 2140, Arica   
316            81        Colegio San Marcos         Santa María 2140, Arica   
...           ...                       ...                             ...   
30184         144  Liceo Bernardo O'Higgins  Eleuterio Ramirez 748, Natales   
30212         144  Liceo Bernardo O'Higgins  Eleuterio Ramirez 748, Natales   
30251         144  Liceo Bernardo O'Higgins  Eleuterio Ramirez 748, Natales   
30290         144  Liceo Bernardo O'Higgins  Eleuterio Ramirez 748, Natales   
30372         144  Liceo Bernardo O'Higgins  Eleuterio Ramirez 748, Natales   

       num_mesa                                     pregunta_glosa  \
1             1  ¿Cuáles son los valores y principios más impor...   
142           5  ¿Cuáles son los valores y principios más impor...   
169           6  ¿Cuáles son los valores y principios más impor...   
283           9  ¿Cuáles son los valores y principios más impor...   
316          10  ¿Cuáles son los valores y principios más impor...   
...         ...                                                ...   
30184         2  ¿Cuáles son los valores y principios más impor...   
30212         3  ¿Cuáles son los valores y principios más impor...   
30251         4  ¿Cuáles son los valores y principios más impor...   
30290         6  ¿Cuáles son los valores y principios más impor...   
30372         9  ¿Cuáles son los valores y principios más impor...   

      concepto_nombre concepto_otro tipo-eleccion  \
1          Democracia           NaN             A   
142        Democracia           NaN             A   
169        Democracia           NaN             P   
283        Democracia           NaN             P   
316        Democracia           NaN             A   
...               ...           ...           ...   
30184      Democracia           NaN             A   
30212      Democracia           NaN             A   
30251      Democracia           NaN             A   
30290      Democracia           NaN             A   
30372      Democracia           NaN             A   

                                              fundamento  len_chars  \
1                                         Sin fundamento         14   
142    Que el ciudadano pueda participar en la toma d...        141   
169    Las condiciones actuales no constituyen democr...        118   
283    por la estructura establecida en el dialogo an...         52   
316    debemos pasar de una democracia representativa...         66   
...                                                  ...        ...   
30184  Incorporarle nuevos elementos de acuerdo a los...        105   
30212  La democracia debe ser la base de la sociedad ...        299   
30251  Descentr

#### Tokenización + POS + Lematización
**Procesamiento con Stanza**  
- Se aplica el pipeline neuronal (`tokenize,pos,lemma`)
- Resultado: lista de tuplas `(token_original, lema, upos)` por documento

In [123]:
def stanza_process(doc_str: str):
    doc = nlp(doc_str)
    rows = []  # (token, lemma, upos)
    for sent in doc.sentences:
        for w in sent.words:
            rows.append((w.text, w.lemma.lower(), w.upos))
    return rows

subset['doc'] = subset['texto_norm'].progress_apply(stanza_process)

100%|██████████| 507/507 [02:30<00:00,  3.36it/s]


#### Visualización de tokens crudos (antes de filtrar stopwords) del subconjunto
- Generar y revisar top 30 de tokens lematizados antes de aplicar stopwords de dominio  

In [124]:
subset['tokens_crudos'] = subset['doc'].apply(lambda toks: [lemma for _, lemma, pos in toks])

freq_crudas = Counter([w for fila in subset['tokens_crudos'] for w in fila])
top30 = freq_crudas.most_common(30)
top30

[('el', 1071),
 ('de', 600),
 ('y', 399),
 ('que', 353),
 ('ser', 328),
 ('democracia', 308),
 ('él', 278),
 ('deber', 232),
 ('uno', 225),
 ('en', 217),
 ('a', 193),
 ('participativo', 181),
 ('participación', 157),
 ('como', 116),
 ('no', 113),
 ('para', 101),
 ('con', 99),
 ('ciudadano', 92),
 ('todo', 90),
 ('representativo', 82),
 ('por', 81),
 ('valor', 77),
 ('vinculante', 72),
 ('pueblo', 65),
 ('fundamental', 60),
 ('estar', 60),
 ('poder', 56),
 ('acuerdo', 55),
 ('derecho', 53),
 ('tener', 52)]

In [ ]:
subset['doc']

#### Stopwords
**Documentación de Stopwords**  
- Documentar la lista de Stopwords en archivo `stopwords_dominio.txt`.

In [125]:
spanish_sw = set(stopwords.words('spanish'))
spanish_sw

{'a',
 'al',
 'algo',
 'algunas',
 'algunos',
 'ante',
 'antes',
 'como',
 'con',
 'contra',
 'cual',
 'cuando',
 'de',
 'del',
 'desde',
 'donde',
 'durante',
 'e',
 'el',
 'ella',
 'ellas',
 'ellos',
 'en',
 'entre',
 'era',
 'erais',
 'eran',
 'eras',
 'eres',
 'es',
 'esa',
 'esas',
 'ese',
 'eso',
 'esos',
 'esta',
 'estaba',
 'estabais',
 'estaban',
 'estabas',
 'estad',
 'estada',
 'estadas',
 'estado',
 'estados',
 'estamos',
 'estando',
 'estar',
 'estaremos',
 'estará',
 'estarán',
 'estarás',
 'estaré',
 'estaréis',
 'estaría',
 'estaríais',
 'estaríamos',
 'estarían',
 'estarías',
 'estas',
 'este',
 'estemos',
 'esto',
 'estos',
 'estoy',
 'estuve',
 'estuviera',
 'estuvierais',
 'estuvieran',
 'estuvieras',
 'estuvieron',
 'estuviese',
 'estuvieseis',
 'estuviesen',
 'estuvieses',
 'estuvimos',
 'estuviste',
 'estuvisteis',
 'estuviéramos',
 'estuviésemos',
 'estuvo',
 'está',
 'estábamos',
 'estáis',
 'están',
 'estás',
 'esté',
 'estéis',
 'estén',
 'estés',
 'fue',
 'f

#### Guardando los stopwords en un archivo de texto


In [126]:
with open('output/tables/stopwords_español.txt', 'w', encoding='utf-8') as f:
    for w in sorted(spanish_sw):
        f.write(w + '\n')
print(f"Stopwords guardadas en el archivo stopwords_español.txt ({len(spanish_sw)} palabras)")


Stopwords guardadas en el archivo stopwords_español.txt (313 palabras)


### **07. Extracción de Sustantivos, Adjetivos y Bigramas**
**Filtrado POS (Part‑Of‑Speech)**  
- Sustantivos y nombres propios (`NOUN`, `PROPN`)  
- Adjetivos (`ADJ`)  
- Bigramas sustantivo–adjetivo contiguos  
- Eliminación de tokens con frecuencia <2 y stopwords de dominio

#### Filtrar Sustantivos

In [127]:
subset['nouns'] = subset['doc'].apply(
    lambda toks: [lemma for _, lemma, pos in toks
                  if pos in {"NOUN","PROPN"} and lemma not in spanish_sw and len(lemma) > 2]
)

In [128]:
subset['nouns']

1                                             [fundamento]
142      [ciudadano, toma, decisión, igualdad, aceptaci...
169      [condición, democracia, obstante, valor, const...
283                                  [estructura, dialogo]
316                                           [democracia]
                               ...                        
30184                          [elemento, acuerdo, tiempo]
30212    [democracia, base, sociedad, modalidad, gobier...
30251    [descentralización, justicia, igualdad, opinió...
30290                                           [concepto]
30372    [valor, sistema, ciudadanía, aumento, priorida...
Name: nouns, Length: 507, dtype: object

#### Filtrar Adjetivos

In [129]:
subset['adjs'] = subset['doc'].apply(
    lambda toks: [lemma for _, lemma, pos in toks
                  if pos == 'ADJ' and len(lemma) > 2]
)

In [130]:
subset['adjs']

1                                                       []
142                                                     []
169                           [actual, fundamental, nuevo]
283                                [establecido, anterior]
316                        [representativo, participativo]
                               ...                        
30184    [nuevo, ademocracia, sustentable, sostenible, ...
30212    [chileno, democratico, fundamental, directo, r...
30251                                                   []
30290                                        [igualitario]
30372                     [necesario, politico, controlar]
Name: adjs, Length: 507, dtype: object

#### Bigramas Sustantivo–Adjetivo contiguos


In [131]:
def noun_adj_bigrams(tok_rows):
    bigs = []
    for i in range(len(tok_rows)-1):
        (_, lemma1, pos1) = tok_rows[i]
        (_, lemma2, pos2) = tok_rows[i+1]
        if pos1 in {'NOUN','PROPN'} and pos2 == 'ADJ':
            if lemma1 not in spanish_sw and lemma2 not in spanish_sw:
                bigs.append(f'{lemma1}_{lemma2}')
    return bigs

subset['bigrams_noun_adj'] = subset['doc'].apply(noun_adj_bigrams)

### **08. Conteos Globales**

**Conteos finales**  
- `all_nouns`: Contador de sustantivos  
- `all_bigrams`: Contador de bigramas N‑A  
- `adj_counts`: Contador de adjetivos  

In [133]:
all_nouns = Counter([w for fila in subset['nouns'] for w in fila])
all_bigrams = Counter([b for fila in subset['bigrams_noun_adj'] for b in fila])
adj_counts = Counter([w for fila in subset['adjs'] for w in fila])
print('Unigramas distintos:', len(all_nouns))
print('Bigramas N-A distintos:', len(all_bigrams))
print('Adjetivos distintos:', len(adj_counts))

Unigramas distintos: 482
Bigramas N-A distintos: 397
Adjetivos distintos: 333


#### Remover claves vacías o sospechosas
Se crea la función `clean_key(k)` para verificar que una palabra no esté vacía, y sea de tipo string. También verifica que no contenga espacios en blancos el string. Se crea un diccionario llamado `filtered_items`, donde se agregan las palabras que pasan el filtro de la función `clean_key(k)` y que aparecerán al menos dos veces en el conteo global de sustantivos.

In [135]:
def clean_key(k):
    return bool(k) and isinstance(k, str) and not k.isspace()

filtered_items = {k:int(v) for k,v in all_nouns.items() if clean_key(k) and v>=2}

print("Vocabulario filtrado:", len(filtered_items), "Total frecuencia:", sum(filtered_items.values()))

Vocabulario filtrado: 213 Total frecuencia: 2339


### **08. Visualizaciones**

#### WordCloud Sustantivos - Concepto Elegido

In [ ]:
freq_unigrams = {w:c for w,c in all_nouns.items() if c>=2}
wc = WordCloud(width=1500, height=800, background_color='white', colormap="magma",  max_words=1000, collocations=False)
wc.generate_from_frequencies(freq_unigrams)
plt.figure(figsize=(12,8))
plt.imshow(wc, interpolation='bilinear')
plt.axis('off')
plt.title(f'WordCloud Sustantivos – {concepto_elegido}')
plt.show()

#### WordCloud Bigramas (Sustantivo-Adjetivo)

In [ ]:
freq_bigrams = {b:c for b,c in all_bigrams.items() if c>=2}
freq_bigrams_space = {b.replace('_', ' '): c for b, c in freq_bigrams.items()} # Se crea un nuevo diccionario reemplazando '_' por ' '

wc2 = WordCloud(
    width=1500,
    height=800,
    background_color='white',
    colormap="magma",
    max_words=400,
    collocations=False
).generate_from_frequencies(freq_bigrams_space)

plt.figure(figsize=(12,8))
plt.imshow(wc2, interpolation='bilinear')
plt.axis('off')
plt.title(f'WordCloud Bigramas Sustantivo - Adjetivo – {concepto_elegido}')
plt.show()

#### Top 20 Adjetivos - Democracia

In [ ]:
# Nuevo dataFrame de adjetivos
adj_df = pd.DataFrame(adj_counts.most_common(20), columns=['adjetivo','frecuencia'])

plt.figure(figsize=(10,6))
bars = plt.barh(adj_df['adjetivo'], adj_df['frecuencia'], color="#ED7272")

# Etiquetas de datos
for bar in bars:
    width = bar.get_width()
    plt.text(
        width + max(adj_df['frecuencia']) * 0.01, 
        bar.get_y() + bar.get_height()/2,
        f'{int(width)}',
        va='center',
        fontsize=9
    )

plt.xlabel('Frecuencia')
plt.ylabel('Adjetivo')
plt.title(f'Top 20 Adjetivos – {concepto_elegido}')
plt.gca().invert_yaxis()
plt.tight_layout()
plt.show()

#### Red de Bigramas

In [ ]:
G = nx.Graph()
for big, freq in freq_bigrams.items():
    noun, adj = big.split('_')
    G.add_node(noun, tipo='NOUN')
    G.add_node(adj,  tipo='ADJ')
    G.add_edge(noun, adj, weight=freq)

print('Nodos:', G.number_of_nodes(), '| Aristas:', G.number_of_edges())

# Cálculo de métricas para tamaño y color
degree_dict = dict(G.degree())

# Se escala el tamaño de nodo según grado
node_sizes  = [100 + degree_dict[n]*200 for n in G.nodes()]
color_map = {'NOUN':"#F88A8A", 'ADJ':"#FFC18B"}
node_colors = [color_map[G.nodes[n]['tipo']] for n in G.nodes()]

# Se escala la anchura de arista según peso 
edge_widths = [G[u][v]['weight']*0.1 for u,v in G.edges()]

plt.figure(figsize=(16,14))
pos = nx.spring_layout(G, k=0.5, seed=42)

nx.draw_networkx_edges(G, pos, width=edge_widths, alpha=0.6)
nx.draw_networkx_nodes(G, pos, node_size=node_sizes, node_color=node_colors, alpha=0.9)
nx.draw_networkx_labels(G, pos, font_size=9)

legend_elems = [
    Patch(facecolor=color_map['NOUN'], label='Sustantivo'),
    Patch(facecolor=color_map['ADJ'],  label='Adjetivo')
]
plt.legend(handles=legend_elems, title='Tipo Nodo')

plt.title(f'Red de Bigramas Sustantivo–Adjetivo – {concepto_elegido}')
plt.axis('off')
plt.tight_layout()
plt.show()